### PR TITLE
Revise the usage of the global error value/exception flag

### LINKED ISSUE #3422: Assertion object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE in ecma_deref_object
###### JerryScript revision
1829d2df

###### Build platform
Linux-4.15.0-72-generic-x86_64-with-Ubuntu-18.04-bionic

###### Build steps
```
./tools/build.py --clean --debug --compile-flag=-fsanitize=address \
--compile-flag=-m32 --compile-flag=-fno-omit-frame-pointer \
--compile-flag=-fno-common --compile-flag=-g \
--strip=off --system-allocator=on --logging=on \
--linker-flag=-fuse-ld=gold --error-messages=on --profile=es2015-subset
```

###### Test case
```javascript
RegExp('{}');
let {t: []} = {t: []}
```

###### Output
```text
Script Error: SyntaxError: RegExp quantifier error: missing digits.
ICE: Assertion 'object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE' failed at jerryscript/jerry-core/ecma/base/ecma-gc.c(ecma_deref_object):146.
Error: ERR_FAILED_INTERNAL_ASSERTION
```

###### Backtrace
```text
bt
#0  0xf7fd5079 in __kernel_vsyscall ()
#1  0xf77fc832 in raise () from /lib/i386-linux-gnu/libc.so.6
#2  0xf77fdcc1 in abort () from /lib/i386-linux-gnu/libc.so.6
#3  0x5657b549 in jerry_port_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-port/default/default-fatal.c:30
#4  0x56644e58 in jerry_fatal (code=ERR_FAILED_INTERNAL_ASSERTION) at jerryscript/jerry-core/jrt/jrt-fatals.c:58
#5  0x56644e99 in jerry_assert_fail (assertion=0x566f03a0 "object_p->type_flags_refs >= ECMA_OBJECT_REF_ONE", file=0x566f0340 "jerryscript/jerry-core/ecma/base/ecma-gc.c", function=0x566a7820 <__func__.5826.lto_priv.895> "ecma_deref_object", line=146) at jerryscript/jerry-core/jrt/jrt-fatals.c:82
#6  0x5669e991 in ecma_deref_object (object_p=0xf5f00760) at jerryscript/jerry-core/ecma/base/ecma-gc.c:146
#7  0x56678990 in ecma_free_value (value=4126148451) at jerryscript/jerry-core/ecma/base/ecma-helpers-value.c:1044
#8  0x56696a95 in jerry_release_value (value=4126148451) at jerryscript/jerry-core/api/jerry.c:1366
#9  0x56691f60 in main (argc=2, argv=0xffffcb24) at jerryscript/jerry-main/main-unix.c:925
```

<sup>Found by [Fuzzinator](http://fuzzinator.readthedocs.io/) with [grammarinator](https://github.com/renatahodovan/grammarinator). </sup>

### PR DESCRIPTION
**Pull request by rerobika**
This patch also fixes #3422.

JerryScript-DCO-1.0-Signed-off-by: Robert Fancsik frobert@inf.u-szeged.hu

### CODE REVIEW

**Comment by dbatyai** [comment]
This has been popping up a lot lately, shouldn't we add a function to correctly retrieve/free the context error value?

**Comment by rerobika** [comment]
@dbatyai Good idea!

File: jerry-core/ecma/builtin-objects/ecma-builtin-promise.c [hunk 1]
```diff
@@ -610,7 +609,7 @@ ecma_builtin_promise_race_or_all (ecma_value_t this_arg, /**< 'this' argument */
 
   if (ECMA_IS_VALUE_ERROR (ret))
   {
-    ret = JERRY_CONTEXT (error_value);
+    ret = jcontext_retrieve_exception ();
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: Why jcontext? Do we have any functions in jcontext? What about `ecma_take_exception` ?

File: jerry-core/ecma/operations/ecma-iterator-object.c [hunk 2]
```diff
@@ -486,7 +487,7 @@ ecma_op_iterator_close (ecma_value_t iterator) /**< iterator value */
     return ECMA_VALUE_UNDEFINED;
   }
 
-  JERRY_CONTEXT (status_flags) |= ECMA_STATUS_EXCEPTION;
+  jcontext_raise_exception (completion);
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-iterator-object.c]: Same thing. Should be an ecma helper.

File: jerry-core/ecma/builtin-objects/ecma-builtin-promise.c (hunk 1 above)
Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: > Why jcontext?

The global error is part of the Jerry context therefore IMO the `jcontext` prefix is more suitable than the `ecma`.

> Do we have any functions in jcontext?
Currently we do not have any functions there, but I think that is not a blocking issue.

Comment (zherczeg) [reply jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: `jcontext_take_exception` ? Its an ownership taking.

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: Sure! I can deal with this renaming.

Comment (zherczeg) [reply jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: `jcontext_catch_exception` could be a good idea.

Comment (rerobika) [reply jerry-core/ecma/builtin-objects/ecma-builtin-promise.c]: I'd prefer the `jcontext_take_exception`.

**Comment by rerobika** [comment]
@zherczeg The patch is updated.

**Pull request merged**